rskj-core/src/main/java/org/ethereum/rpc/LogFilter.java

```diff
@@ -231,47 +232,75 @@ private static void processSingleBlockByHash(String blockHash, Blockchain blockc
             return;
         }
 
-        long blockNumber = blockByHash.getNumber();
-        processBlocks(blockNumber, blockNumber, filter, blockchain, blocksBloomStore);
+        processBlocks(blockByHash, blockByHash, filter, blockchain, blocksBloomStore);
     }
 
-    private static void processBlocks(long fromBlockNumber, long toBlockNumber, LogFilter filter, Blockchain blockchain, BlocksBloomStore blocksBloomStore) {
-        BlocksBloom auxiliaryBlocksBloom = null;
-        long bestBlockNumber = blockchain.getBestBlock().getNumber();
+    private static void processBlocks(Block fromBlock, Block toBlock, LogFilter filter, Blockchain blockchain, BlocksBloomStore blocksBloomStore) {
+        final long bestBlockNumber = blockchain.getBestBlock().getNumber();
 
-        for (long blockNum = fromBlockNumber; blockNum <= toBlockNumber; blockNum++) {
-            boolean isConfirmedBlock = blockNum <= bestBlockNumber - blocksBloomStore.getNoConfirmations();
+        BlocksBloom bloomAccumulator = null;
 
-            if (isConfirmedBlock) {
-                if (blocksBloomStore.firstNumberInRange(blockNum) == blockNum) {
-                    if (blocksBloomStore.hasBlockNumber(blockNum)) {
-                        BlocksBloom blocksBloom = blocksBloomStore.getBlocksBloomByNumber(blockNum);
+        Block block = toBlock;
+        while (block != null && block.getNumber() >= fromBlock.getNumber()) {
+            boolean isConfirmedBlock = block.getNumber() <= bestBlockNumber - blocksBloomStore.getNoConfirmations();
 
-                        if (!filter.addressesTopicsFilter.matchBloom(blocksBloom.getBloom())) {
-                            blockNum = blocksBloomStore.lastNumberInRange(blockNum);
-                            continue;
-                        }
-                    }
+            BlocksBloom blocksBloom = isConfirmedBlock ? blocksBloomStore.getBlocksBloomByNumber(block.getNumber()) : null;
+            boolean hasBloom = blocksBloom != null;
 
-                    auxiliaryBlocksBloom = new BlocksBloom();
-                }
+            if (hasBloom && !bloomMatches(blocksBloom, filter)) {
+                block = getBlockToSkipToOrNullIfDone(block, fromBlock.getNumber(), blocksBloomStore, blockchain);
+                continue;
+            }
 
-                Block block = blockchain.getBlockByNumber(blockNum);
+            boolean needsBloom = isConfirmedBlock && !hasBloom;
+            if (needsBloom) {
+                bloomAccumulator = addBlockBloom(block, bloomAccumulator, blocksBloomStore);
+            }
 
-                if (auxiliaryBlocksBloom != null) {
-                    auxiliaryBlocksBloom.addBlockBloom(blockNum, new Bloom(block.getLogBloom()));
-                }
+            // reverseTxOrder=true to process txs in reverse order for later complete list reverse
+            filter.onBlock(block, true);
+            block = blockchain.getBlockByHash(block.getParentHash().getBytes());

```

Comment:
What is the advantage of work with the block object instead of the block number into the loop?

Reply:
We do less queries compared to `blockStore#getChainBlockByNumber` and also when using hash and looping backwards we prevent a potential bug on forks. @Vovchyk spotted this 2 things.